core/xml.js

```diff
@@ -631,125 +634,164 @@ Blockly.Xml.domToBlockHeadless_ = function(xmlBlock, workspace) {
   var id = xmlBlock.getAttribute('id');
   block = workspace.newBlock(prototypeName, id);
 
-  var blockChild = null;
+  // Preprocess childNodes so tags can be processed in a consistent order.
+  /**
+   * A mapping of nodeName to node for child nodes of xmlBlock.
+   * @type {{next: Array<Element>, mutation: Array<Element>,
+   *    input: Array<Element>, data: Array<Element>, field: Array<Element>,
+   *    comment: Array<Element>}}
+   */
+  var xmlChildNameMap = {
+    mutation: [], comment: [], data: [], field: [], input: [],
+    next: []
+  };
   for (var i = 0, xmlChild; (xmlChild = xmlBlock.childNodes[i]); i++) {
```

Comment:
This loop can all be pulled out into a helper function that returns the generated map.

Reply:
Done.